### PR TITLE
* Sponge: Add water absorbtion

### LINKED ISSUE #1411: Add sponge interaction
http://minecraft.gamepedia.com/Sponge

The sponge should remove water in its vicinity.

### PR DESCRIPTION
**Pull request by earboxer**
closes #1411

### CODE REVIEW

File: src/Blocks/BlockSponge.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	bool GetSoaked(Vector3i a_Rel, cChunk & a_Chunk)
+	{
+		static const std::array<Vector3i, 5> WaterCheck
+		{
+			{
+				{ 1, 0,  0},
+				{-1, 0,  0},
+				{ 0, 0,  1},
+				{ 0, 0, -1},
+				{ 0, 1,  0},
```
Comment (peterbell10) [inline src/Blocks/BlockSponge.h]: This seems to be a deliberate choice to ignore the -Y neighbour but when I try it in 1.12, it definitely soaks. Even if it's just a single water block that's below it.

File: src/Blocks/BlockSponge.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			}
+		};
+		BLOCKTYPE TargetBlock;
+		NIBBLETYPE TargetMeta;
+		a_Chunk.GetBlockTypeMeta(a_Rel.x, a_Rel.y, a_Rel.z, TargetBlock, TargetMeta);
+		if (TargetMeta == E_META_SPONGE_DRY){
+		bool ShouldSoak = std::any_of(WaterCheck.cbegin(), WaterCheck.cend(), [a_Rel, & a_Chunk](Vector3i a_Offset)
+			{
+				BLOCKTYPE NeighborType;
+				return (
```
Comment (peterbell10) [inline src/Blocks/BlockSponge.h]:
```suggestion
				return IsWet(a_Rel + a_Offset, a_Chunk);
```

File: src/Blocks/BlockSponge.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+				{ 1, 0,  0},
+				{-1, 0,  0},
+				{ 0, 0,  1},
+				{ 0, 0, -1},
+				{ 0, 1,  0},
+			}
+		};
+		BLOCKTYPE TargetBlock;
+		NIBBLETYPE TargetMeta;
+		a_Chunk.GetBlockTypeMeta(a_Rel.x, a_Rel.y, a_Rel.z, TargetBlock, TargetMeta);
```
Comment (peterbell10) [inline src/Blocks/BlockSponge.h]: If you only need the meta, you can simplify this
```suggestion
		NIBBLETYPE TargetMeta = a_Chunk.GetBlockMeta(a_Rel);
```

File: src/Blocks/BlockSponge.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+				return (
+					a_Chunk.UnboundedRelGetBlockType(a_Rel.x + a_Offset.x, a_Rel.y + a_Offset.y, a_Rel.z + a_Offset.z, NeighborType)
+					&& IsBlockWater(NeighborType)
+				);
+			}
+		);
+
+		if (ShouldSoak)
+		{
+			// TODO: Keep track of depth, don't go more than 7 blocks away
```
Comment (peterbell10) [inline src/Blocks/BlockSponge.h]: I think you just need to change the queue to
```cpp
struct sSeed
{
	Vector3i m_Pos;
	int m_Depth;
};
std::queue<sSeed> Seeds;
```

File: src/Blocks/BlockSponge.h (hunk 2 above)
Comment (peterbell10) [reply src/Blocks/BlockSponge.h]: You can either make `IsWet` static or explicitly add `this` to the lambda capture list.
```cpp
[this, a_Rel, & a_Chunk](Vectir3i Offset) { return IsWet(a_Rel + a_Offset, a_Chunk); };
```

**Comment by earboxer** [comment]
Fixed the chunk boundary issue.

File: src/Blocks/BlockSponge.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+			{
+				std::queue<sSeed> Seeds;
+				int count = 0;
+				const int maxDepth = 7;
+				Seeds.emplace(a_Rel + WaterCheck[0], maxDepth - 1);
+				Seeds.emplace(a_Rel + WaterCheck[1], maxDepth - 1);
+				Seeds.emplace(a_Rel + WaterCheck[2], maxDepth - 1);
+				Seeds.emplace(a_Rel + WaterCheck[3], maxDepth - 1);
+				Seeds.emplace(a_Rel + WaterCheck[4], maxDepth - 1);
+				Seeds.emplace(a_Rel + WaterCheck[5], maxDepth - 1);
```
Comment (peterbell10) [inline src/Blocks/BlockSponge.h]: This looks like a loop to me!

File: Server/Plugins/APIDump/APIDesc.lua [hunk 6]
```diff
@@ -16582,6 +16582,14 @@ end
 				{
 					Notes = "A flag in the metadata of heads that indicates that the head is a zombie head.",
 				},
+				E_META_SPONGE_DRY =
+				{
+					Notes = "A flag in the metadata of sponges that indicates that the sponge is dry.",
+				}
```
Comment (peterbell10) [inline Server/Plugins/APIDump/APIDesc.lua]: Missing comma

File: src/Blocks/BlockSponge.h [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		cBlockHandler::Check(a_ChunkInterface, a_PluginInterface, a_RelX, a_RelY, a_RelZ, a_Chunk);
+	}
+
+
+
+
+
+	/** Check blocks above and around to see if they are water.
+		If one is, soak this sponge and remove nearby water.
+		Returns TRUE if the block was changed. */
```
Comment (madmaxoft) [inline src/Blocks/BlockSponge.h]: The doxy-comment should be indented all at the same level.

File: src/Blocks/BlockSponge.h [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		bool ShouldSoak = std::any_of(WaterCheck.cbegin(), WaterCheck.cend(), [a_Rel, & a_Chunk](Vector3i a_Offset)
+			{
+				return IsWet(a_Rel + a_Offset, a_Chunk);
+			}
+		);
+
+		if (! ShouldSoak)
+		{
+			return false;
+		}
```
Comment (madmaxoft) [inline src/Blocks/BlockSponge.h]: Add comments about what the code is doing, in groups.

Group lines 62 to 77 into a single block, with comment along the lines of `Check if this is a dry sponge next to a water block`

File: src/Blocks/BlockSponge.h [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+					}
+				}
+			}
+			Seeds.pop();
+		}
+		a_Chunk.SetBlock(a_Rel.x, a_Rel.y, a_Rel.z, E_BLOCK_SPONGE, E_META_SPONGE_WET);
+		return true;
+	}
+
+	void DryUp(Vector3i a_Rel, cChunk & a_Chunk)
```
Comment (madmaxoft) [inline src/Blocks/BlockSponge.h]: This could be `static`.

Style: 5 lines between functions.

**Pull request merged**